### ISSUE TITLE
Userland 1 - Plugins, SPAs, Node, or?

### ISSUE THREAD

**Issue by pfraze** [ssbc/phoenix#370]
Off the discussion in #368, I want to pause briefly and ask what kind of userland interests us as app developers?

(For brevity, I ask that you read through #368 to get the full background.)

The goal is to deploy software to each other, to our communities, to our teams, to strangers for fun (viral, virus, what's the difference?). The software will integrate, at minimum through the shared feed.

Would it be more fun to use node as the environment? We'd publish packages, and sbot would install to a sub of the home directory. If it's a service, sbot would orchestrate it. If it's a web app, it would run on a localhost port, and phoenix would link to it (probably like how google links to its other properties from gmail).

OR

Would it be more fun to deploy SPAs? Sbot would host the SPAs on their own pages, and phoenix would link out to them, same as with node.

OR

Would it be more fun to deploy phoenix plugins? The packages would be evaled in the page at startup. Phoenix would provide an API to hook into the UI with new behaviors, and to add new pages.

OR

(your suggestion here?)

**Comment by No9** [comment]
Thanks for opening this @pfraze 

**TL;DR;** 
1. sbot should be more isolated from phoenix. 
2. Phoenix should focus on being a discussion app rather than an application development platform.
3. sbot should be API and services. 

Apologies if the remarks that follow have been discussed before or move away from the core ideals of SSBC but this is how I see things after a couple of months mooching around.

Phoenix is acting as a fantastic reference implementation of the features in scuttlebot and while it is the primary use case of sbot for now I think the current coupling will prevent the reuse of the core application infrastructure in wider scenarios. [i.e. this list of suggested applications](https://github.com/ssbc/scuttlebot/issues/146#issuecomment-93092946)
I can't see these applications starting from phoenix as a base. 

To use the analogy in https://github.com/ssbc/phoenix/pull/368#issuecomment-92554865
sbot === git 
i.e. a well documented set of api's and command-line tools 

phoenix === github 
i.e. An application of sbot

sbot should be open to the possibilities completely different applications as git has gitlab, bitbucket etc. 

If sbot took that approach it will probably make [suggestions like having a templating system ](https://github.com/ssbc/scuttlebot/issues/127) easier as the whole UI stack can be replaced if necessary. 

To elaborate on the points made by @ahdinosaur https://github.com/ssbc/phoenix/pull/368#issuecomment-92534777 

1. try to be a sweet mass-market discussion app
I think phoenix should continue with this goal but the coupling with sbot needs to be made looser.

2. try to be a sweet raw message app for ssb developers
As above in the git analogy sbot should be a set of apis. Phoenix shouldn't try and solve this. 

3. try to be a sweet generalized platform for hosting ssb apps
I share the concerns about this 

Based/biased on the experiences I had on http://lockerproject.org
(A social feed aggregator with the idea of sharing these feeds using telehash http://telehash.org/)

Lockerproject had the idea of [applets](https://github.com/LockerProject/Locker/tree/master/Apps) that were node_modules.

This approach caused applications to be "constrained" by the need to be hosted on a dashboard/entry point that had to align to the overall look and feel, otherwise they would feel out of place and make for a poor user experience.
It also mandated certain development approaches/libraries which in turn caused "discussion/bikeshedding"

Eventually Locker was dropped for a more API based approach in [hallway](https://github.com/singly/hallway) Hence the push to not only focus on phoenix as *the* only implementation in town. 

Hope this is helpful in some way.

**Comment by No9** [comment]
**TL;DR; should we move phoenix-api back into sbot**
Went a little further with this today

I think the outcome of my suggestion would be that 
https://www.npmjs.com/package/phoenix-api should be in sbot or more aligned to it 
i.e. sbot-api?
That matches the level of abstraction I was trying to communicate above.

But looking at this issue [is scuttlebot bundled with phoenix](https://github.com/ssbc/phoenix/issues/321) it states there is aready an sbot API?

This seems to be more around plugin development (manifest and RPC Events) as phoenix-api seems to hook directly onto the DB layer.

**Comment by pfraze** [comment]
Those are some good thoughts, @No9. Interested to hear other people's thoughts.

Phoenix-api is part of the sbot api, it just lives in another module/repo than sbot for historical reasons. We may move some of its functions (eg around names and profiles) into another sbot plugin, though -- the phoenix plugin is a bit broad.

**Comment by karissa** [comment]
I think when I was first thinking 'application' I was thinking that pheonix would be an example of these applications. I agree that they should be decoupled significantly, so that developers can have freedom to develop entirely separate user interfaces than pheonix.

**Comment by pfraze** [comment]
I suggest we document the usage-pattern for desktop applications using the master keys, which is:

 - Install sbot and start the server
 - Create/install your application locally
 - The application loads `~/.ssb/private` and connects using https://github.com/ssbc/ssb-client, giving full access rights

That'll get third-party applications going without waiting for the ssb package manager.

**Comment by clehner** [comment]
I like the idea of plugins for message types in phoenix as we've been exploring. I'm also interested in seeing applications other than phoenix running on ssb. @pfraze would applications need to authenticate using the private key if they are just going to connect to localhost, i.e. how I understand phoenix does it now?

**Comment by pfraze** [comment]
@clehner for authenticating without the private key, there are two schemes at our disposal here.

 1. The app provides its own keypair, in which case it's free to sign new messages on its own feed and send them to the local sbot. This is implemented now; if you make a connection to sbot from localhost and auth with a non-master keypair, you'll get `add` rights so you can do this (but no `publish` rights, which you'd need to publish to the master feed).

 2. The app asks for an access token, in an OAuth sort of flow. Phoenix previously supported this, but it was removed because it only worked for web applications, and we're not sure about web apps due to their security profile.

**Comment by No9** [comment]
Ah @pfraze thanks for the heads up on ssb-client
